### PR TITLE
Add a GeocentricSolarEcliptic Frame 

### PR DESCRIPTION
**Pull request by shreyasbapat**
Addresses #335

### CODE REVIEW

**Comment by shreyasbapat** [comment]
After thinking about it! I have some questions in mind! 
In @Juanlu001  's pull request https://github.com/astropy/astropy/pull/6508 he has introduced a transformation matrix which is basically for `TrueEcliptic` or `MeanEcliptic` (That is still not clear). But transformation to GSE is very easy in way that Geocentric True Ecliptic and Geocentric solar ecliptic, both share their X-Y Axes in ecliptic plane. I just have to rotate the coordinate axes by an angle which is **angle between earth-sun and earth-equinox(true)**. 
First, how do I get that angle? 
If I don't have to get that angle, from where do I get the transformation matrix? Does ERFA has such a matrix? I assume no?

**Comment by Juanlu001** [comment]
Okay, now I understand the problem.

> Does ERFA has such a matrix? I assume no?

I am not sure. I looked at it very quickly and didn't see it. We should look more closely.

> If I don't have to get that angle, from where do I get the transformation matrix?

Even if it's not in ERFA, I see some formulas here and there that use the "Sun ecliptic longitude":

https://www.spenvis.oma.be/help/background/coortran/coortran.html#T2

However, a random website is not enough and I'd prefer to look at them from some source.

**Comment by shreyasbapat** [comment]
Oh! GEI is the same as Geocentric True Ecliptic! I also saw this!

**Comment by Juanlu001** [comment]
This will need a rebase now

**Comment by shreyasbapat** [comment]
Sorry for a looooong break. I will start working on it

**Comment by shreyasbapat** [comment]
> https://www.spenvis.oma.be/help/background/coortran/coortran.html#T2

Should I go with this?

**Comment by Juanlu001** [comment]
Take a look at https://docs.sunpy.org/en/stable/api/sunpy.coordinates.transformations.hcrs_to_hgs.html and https://github.com/sunpy/sunpy/blob/master/sunpy/coordinates/transformations.py#L304 first, and try to understand if we can reuse part of that transformation. Notice that we don't intend to depend on SunPy though.

**Comment by shreyasbapat** [comment]
This is helpful!

**Comment by shreyasbapat** [comment]
Back to it finally! Rebased and started exploring sunpy links!

File: src/poliastro/frames.py [hunk 1]
```diff
@@ -220,6 +222,78 @@ class PlutoICRS(_PlanetaryICRS):
...
+    rotation_angle = -np.arccos(
+        A.dot(B) / (A.norm() * B.norm())
+    )  # negation is required
+
+    # This line works around some input/output quirks of Astropy's rotation_matrix()
+    matrix = np.array(rotation_matrix(rotation_angle, rotation_axis.xyz.value.tolist()))
+    return matrix
+
+
+_EARTH_NORTH_POLE_GCRS = UnitSphericalRepresentation(lon=0 * u.deg, lat=90 * u.deg)
```
Comment (shreyasbapat) [inline src/poliastro/frames.py]: I have some doubt here! 

Should this be: 

```
_EARTH_NORTH_POLE_GCRS = UnitSphericalRepresentation(lon=0*u.deg, lat=-66.5*u.deg)

_EARTH_DETILT_MATRIX = _make_rotation_matrix_from_reprs(_EARTH_NORTH_POLE_GCRS, CartesianRepresentation(0, 0, 1))

```
This is taking the angle from the axis that's perpendicular to orbit. But I think the first one is correct. Doing this introduces some negetive signs in the matrix.

Comment (shreyasbapat) [reply src/poliastro/frames.py]: 66.5 deg is 90- 23.5 (Tilt of the Earth Axis)

File: src/poliastro/frames.py [hunk 2]
```diff
@@ -220,6 +226,76 @@ class PlutoICRS(_PlanetaryICRS):
...
+    rotation_angle = -np.arccos(
+        A.dot(B) / (A.norm() * B.norm())
+    )  # negation is required
+
+    # This line works around some input/output quirks of Astropy's rotation_matrix()
+    matrix = np.array(rotation_matrix(rotation_angle, rotation_axis.xyz.value.tolist()))
+    return matrix
+
+
+_EARTH_NORTH_POLE_GCRS = UnitSphericalRepresentation(lon=0 * u.deg, lat=66.5 * u.deg)
```
Comment (shreyasbapat) [inline src/poliastro/frames.py]: I figured out what is correct. Can you please verify this?

File: src/poliastro/frames.py [hunk 3]
```diff
@@ -220,6 +226,78 @@ class PlutoICRS(_PlanetaryICRS):
...
+@frame_transform_graph.transform(DynamicMatrixTransform, GCRS, GeocentricSolarEcliptic)
+def gcrs_to_geosolarecliptic(gcrs_coo, to_frame):
+
+    if to_frame.obstime is None:
+        raise ValueError(
+            "To perform this transformation the coordinate"
+            " Frame needs an obstime Attribute"
+        )
+
+    sun_pos_icrs = get_body_barycentric("sun", to_frame.obstime)
```
Comment (shreyasbapat) [inline src/poliastro/frames.py]: And would changing the frame of this matter? As we are finding the difference vector

File: src/poliastro/frames.py [hunk 4]
```diff
@@ -220,6 +227,78 @@ class PlutoICRS(_PlanetaryICRS):
...
+        A.dot(B) / (A.norm() * B.norm())
+    )  # negation is required
+
+    # This line works around some input/output quirks of Astropy's rotation_matrix()
+    matrix = np.array(rotation_matrix(rotation_angle, rotation_axis.xyz.value.tolist()))
+    return matrix
+
+
+_EARTH_ORBIT_PERPEN_POINT_GCRS = UnitSphericalRepresentation(
+    lon=0 * u.deg, lat=66.5 * u.deg
```
Comment (Juanlu001) [inline src/poliastro/frames.py]: This is called the "obliquity" and can be obtained from ERFA, see for example 

https://github.com/astropy/astropy/blob/v3.1.2/astropy/coordinates/builtin_frames/ecliptic_transforms.py#L22-L33

File: src/poliastro/frames.py [hunk 5]
```diff
@@ -220,6 +227,78 @@ class PlutoICRS(_PlanetaryICRS):
...
+            " Frame needs an obstime Attribute"
+        )
+
+    sun_pos_gcrs = get_body("sun", to_frame.obstime).cartesian
+    earth_pos_gcrs = get_body("earth", to_frame.obstime).cartesian
+    sun_earth = sun_pos_gcrs - earth_pos_gcrs
+
+    sun_earth_detilt = sun_earth.transform(_EARTH_DETILT_MATRIX)
+
+    x_axis = CartesianRepresentation(1, 0, 0)
```
Comment (Juanlu001) [inline src/poliastro/frames.py]: What does this x axis represent, and in which reference frame is it written?

File: src/poliastro/frames.py [hunk 6]
```diff
@@ -220,6 +227,78 @@ class PlutoICRS(_PlanetaryICRS):
...
+        raise ValueError(
+            "To perform this transformation the coordinate"
+            " Frame needs an obstime Attribute"
+        )
+
+    sun_pos_gcrs = get_body("sun", to_frame.obstime).cartesian
+    earth_pos_gcrs = get_body("earth", to_frame.obstime).cartesian
+    sun_earth = sun_pos_gcrs - earth_pos_gcrs
+
+    sun_earth_detilt = sun_earth.transform(_EARTH_DETILT_MATRIX)
```
Comment (Juanlu001) [inline src/poliastro/frames.py]: Isn't this essentially ecliptic coordinates? If I'm right, then this transformation could be made simpler by connecting it to 

https://github.com/poliastro/poliastro/blob/c8f316c362da7be277c08397b173a1f65009deab/src/poliastro/frames.py#L49-L58

File: src/poliastro/frames.py (hunk 4 above)
Comment (shreyasbapat) [reply src/poliastro/frames.py]: So, I will try obtaining it fro ERFA!

File: src/poliastro/frames.py (hunk 5 above)
Comment (shreyasbapat) [reply src/poliastro/frames.py]: This represents the earth-sun line in GSE

File: src/poliastro/frames.py (hunk 6 above)
Comment (shreyasbapat) [reply src/poliastro/frames.py]: I will have a look asap

File: src/poliastro/frames.py (hunk 5 above)
Comment (Juanlu001) [reply src/poliastro/frames.py]: Sure, let's add a code comment then :)

Comment (shreyasbapat) [reply src/poliastro/frames.py]: Sure!

File: src/poliastro/frames.py (hunk 6 above)
Comment (shreyasbapat) [reply src/poliastro/frames.py]: I did the GCRS to GSE transformation, and I guess it's easier

File: src/poliastro/frames.py (hunk 5 above)
Comment (Juanlu001) [reply src/poliastro/frames.py]: Shouldn't the comment be _above_ the line? :)

**Comment by shreyasbapat** [comment]
Sure thing. Will do it tonight!

**Comment by shreyasbapat** [comment]
Codeclimate will never pass :/ We will have to supress it.

**Comment by Juanlu001** [comment]
According to Codecov, there are two untested branches:

https://codecov.io/gh/poliastro/poliastro/compare/e1b893c11764d301f584071cd50573e9a6947e93...c32b516398de55d9008c1c7d057ef3774c0e7ce1/diff

* `obstime=None`
* `obstime` not scalar

Please? :innocent:

File: src/poliastro/tests/test_frames.py [hunk 7]
```diff
@@ -109,3 +112,20 @@ def test_icrs_body_position_to_planetary_frame_yields_zeros(body, frame):
         )
 
     assert_quantity_allclose(vector_result.xyz, [0, 0, 0] * u.km, atol=1e-7 * u.km)
+
+
+def test_round_trip_from_GeocentricSolarEcliptic_gives_same_results():
+    gcrs = SkyCoord("02h31m49.09s", "+89d15m50.8s", frame=GCRS(obstime=Time("J2000")))
```
Comment (Juanlu001) [inline src/poliastro/tests/test_frames.py]: And also, I prefer that we avoid the use of `SkyCoord`, if possible.

Comment (shreyasbapat) [reply src/poliastro/tests/test_frames.py]: Done :)

**Comment by shreyasbapat** [comment]
> According to Codecov, there are two untested branches:
https://codecov.io/gh/poliastro/poliastro/compare/e1b893c11764d301f584071cd50573e9a6947e93...c32b516398de55d9008c1c7d057ef3774c0e7ce1/diff
obstime=None
obstime not scalar
Please?

Just realised, astropy infrastructure doesn't allow us to pass `None` anyways. It defaults to J2000 automatically. And neither does it allow to pass list.

**Comment by Juanlu001** [comment]
:heart_eyes: Thanks a lot!

**Pull request merged**